sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3181,8 +3184,12 @@ pi_result piEventsWait(pi_uint32 NumEvents, const pi_event *EventList) {
       // Event has been signaled: If the fence for the associated command list
       // is signalled, then reset the fence and command list and add them to the
       // available list for reuse in PI calls.
-      if (EventList[I]->Queue->RefCount > 0) {
-        EventList[I]->Queue->ZeCommandListFenceMapMutex.lock();
+      auto Queue = EventList[I]->Queue;
+
+      if (Queue->RefCount > 0) {
```

Comment:
I wonder if really need this now that we lock the whole Queue

Reply:
I think that we do.  We only lock the queue while a call to one of the pi interfaces is using the queue.  The retain/release calls still need to be used to reference count the queue itself to keep a queue from being deleted while it was still "in use" by another thread.  say the execution goes like this:  t1 - retain, t1 piEnqueBegin q locked, t1 EnqueEnd q unlocked, t2 - retain, t2 enqueBegin q locked, t2 end q unlocked, t2 release, t1 release.  That first call from t2 to release cannot get rid of the queue, or undo any of it's data structures.

Reply:
Right. Shouldn't the Queue lock be acquired before accessing its RefCount then?

Reply:
No, RefCount is atomic, and is only used to make sure the _pi_queue doesn't get deleted/freed before all threads are finished with it.  You cannot lock the queue before checking it's reference count.  RefCount checking is the first thing that must happen.

Reply:
> You cannot lock the queue before checking it's reference count.

Why?

Reply:
Hmm.  Thinking on it more, you are probably right.  The code probably needs to lock before incrementing or decrementing RefCount, and before reading RefCount when checking for RefCount > 0.  I'll update the code.